pylanetary/spectral/core.py

```diff
@@ -0,0 +1,808 @@
+    
+    '''
+    
+    pars = model[fit_type]['lmfit_func'].guess(self.spec,x=self.xaxis)
+    fit_result = model[fit_type]['lmfit_func'].fit(self.spec,pars,x=self.xaxis,scale_covar=False)
+    print(fit_result.fit_report())
+    save_modelresult(fit_result, outfile+'.sav')
+    
+  ###############################################################################
+    
+  def fit_profile(self, fit_type, showplots = True, plot_name = None, initial_fit_guess = None):
+    '''
+    Calculate the best fit of a Spectral object using the LMFIT algorithm.
+    
+    Parameters
+    ----------
+    fit_type : string
+          The type of line shape used to fit the spectra, string. Options are:
+            All: Fit the spectrum with all line 5 line shape options. Data is always plotted.
+            Gaussian: fit the spectrum with a Gaussian function
+            Lorentzian: fit the spectrum with a Lorentzian function
+            Voigt: fit the spectrum with a Voigt function
+            PseudoV: fit the spectrum with a Pseudo-Voigt function
+            Moffat: fit the spectrum with a Moffat function
+            A description of all line shapes can be found at https://lmfit.github.io/lmfit-py/builtin_models.html
+    showplots : bool, optional
+          The indicator of whether to plot the best fit line shape or a comparison of line shapes. Default is True, which produces plots.
+    plot_name : string, optional
+          The name of a saved figure. Default is None, which does not save a plot.
+    initial_fit_guess: string, optional.
+          The path to a saved LMFIT result to use as the initial guess. Default is None; the LMFIT .guess method is used. 
+    
+    Returns
+    -------
+    peak : float
+          The frequency or wavelength value of the line peak, which is LMFIT's 'center' parameter.
+    err : float
+          The standard deviation of the peak frequency or wavelength value.
+    chi :
+          The calculated Chi-squared statistic of the line fit.
+    chi_red :
+          The calculated reduced Chi-squared statistic of the line fit.
+    
+    Notes
+    -----
+    Guide to keys in the nested model dictionary:
+          model[] : 
+                The Model name, taken from the user input.
+          ref : 
+                A short reference name of model type.
+          lmfit_func: 
+                The built-in LMFIT function name, needed to properly interface with the LMFIT package.
+          label :
+                The label for the legend included in the final plots.
+          color : 
+                The color of the line for a given fit type when plotting.
+    
+    An initial fit guess is not necessary, but can be very beneficial if the spectra are noisy. If your fit is failing, try including an intial guess to improve the performance of LMFIT. It is recommended that you use a pixel with as small a Doppler shift as possible to make your initial guess save file.
+    
+    '''
+    
+    if fit_type != 'All':
```

Comment:
This looks much more modular than before, nice job.  However, looking at this again I am not sure fit_type == 'all' should be supported by this function.  I think it would potentially be much cleaner to have this function handle fitting and plotting a single fit instance, and then show in the documentation or tutorials how to easily create and plot all of them.

The main reason for this is that it removes the "special treatment" that fitting "all" requires, in particular the fact that if fit_type == "all" the returns of this function aren't very logical

It might be good for us to meet again and write down some pseudocode for this, particularly where it relates to plotting.